crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -168,11 +168,26 @@ use crate::Db;
 /// resulting from a few files with a lot of boolean expressions and `if`-statements.
 const MAX_RECURSION_DEPTH: usize = 24;
 
+/// A ternary formula that defines under what conditions a binding is visible. (A ternary formula
+/// is just like a boolean formula, but with `Ambiguous` as a third potential result. See the
+/// module documentation for more details.)
+///
+/// The primitive atoms of the formula are [`Constraint`]s, which express some property of the
+/// runtime state of the code that we are analyzing.
+///
+/// We assume that each atom has a stable value each time that the formula is evaluated. An atom
+/// that resolves to `Ambiguous` might be true or false, and we can't tell which — but within that
+/// evaluation, we assume that the atom has the _same_ unknown value each time it appears. That
+/// allows us to perform simplifications like `A ∨ !A → true` and `A ∧ !A → false`.
+///
+/// That means that when you are constructing a formula, you might need to create distinct atoms
+/// for a particular [`Constraint`], if your formula needs to consider how a particular runtime
+/// property might be different at different points in the execution of the program.
```

Comment:
It's late here already, but just reading this, I wonder why it applies to while conditions, but not to something like
```py
for _ in iterable:
    if <cond>:
        # …
```
Is it because we merge control flow twice here (after the if, after the for) instead of just once in the while loop? Would a `break` point inside the `if` complicate the situation?

Reply:
I think it's because currently an `if` test condition generates no visibility constraints. (We don't attempt to detect empty iterables.) If we did add that feature, we might need to consider this, but I'm still not sure it would apply unless we also tried to detect infinite iterables! Which I doubt we would ever do.

Reply:
Not sure I've totally thought through all the possible concrete visibility constraint trees we might generate here, but my intuition says it's not an issue in this scenario, because there is no difference between any one iteration of the loop and any other one, in terms of the effect on which code is visible or not. Which is not true in a `while` loop, where the key fact is that a static `True` means we'll never exit the loop, a static `False` means we'll never enter it, and a `True` that later changes to `False` means we can enter and then exit.

Reply:
It seems to me like the following two are similar control flow constructs (ignoring `else` branches in loops…):
```py
while cond:
    # <code block>
```
and
```py
for _ in itertools.repeat(None):
    if not cond:
        break
    # <code block>
```


It looks like we can handle the former, but not the latter:
```py
import itertools

cond = False

w = 1
f = 1

while cond:
    w = 2

for _ in itertools.repeat(None):
    if not cond:
        break

    f = 2

print(reveal_type(w))  # Literal[1]
print(reveal_type(f))  # Literal[1, 2] ?
```

Or am I missing something? This might not be related to this PR at all(?).